examples/src/main/java/com/mapbox/navigation/examples/core/ReRouteActivity.kt

```diff
@@ -0,0 +1,300 @@
+import com.mapbox.navigation.base.trip.model.RouteProgress
+import com.mapbox.navigation.core.MapboxNavigation
+import com.mapbox.navigation.core.directions.session.RoutesObserver
+import com.mapbox.navigation.core.directions.session.RoutesRequestCallback
+import com.mapbox.navigation.core.trip.session.LocationObserver
+import com.mapbox.navigation.core.trip.session.RouteProgressObserver
+import com.mapbox.navigation.examples.R
+import com.mapbox.navigation.examples.utils.Utils
+import com.mapbox.navigation.examples.utils.extensions.toPoint
+import com.mapbox.navigation.ui.camera.NavigationCamera
+import com.mapbox.navigation.ui.map.NavigationMapboxMap
+import com.mapbox.navigation.ui.map.NavigationMapboxMapInstanceState
+import kotlinx.android.synthetic.main.activity_reroute_layout.*
+import kotlinx.android.synthetic.main.activity_trip_service.mapView
+import timber.log.Timber
+
+class ReRouteActivity : AppCompatActivity(), OnMapReadyCallback, PermissionsListener {
+
+    companion object {
+        const val MAP_INSTANCE_STATE_KEY = "navgation_mapbox_map_instance_state"
+        const val DEFAULT_INTERVAL_IN_MILLISECONDS = 1000L
+        const val DEFAULT_MAX_WAIT_TIME = DEFAULT_INTERVAL_IN_MILLISECONDS * 5
+    }
+
+    private var permissionsManager: PermissionsManager? = null
+    private var locationEngine: LocationEngine? = null
+    private var mapboxNavigation: MapboxNavigation? = null
+    private var navigationMapboxMap: NavigationMapboxMap? = null
+    private var mapInstanceState: NavigationMapboxMapInstanceState? = null
+    private var destination: LatLng? = null
+    private var directionsRoute: DirectionsRoute? = null
+
+    @SuppressLint("MissingPermission")
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_reroute_layout)
+        mapView.onCreate(savedInstanceState)
+        mapView.getMapAsync(this)
+
+        val mapboxNavigationOptions = MapboxNavigation.defaultNavigationOptions(
+                this,
+                Utils.getMapboxAccessToken(this)
+        )
+
+        mapboxNavigation = MapboxNavigation(
+                applicationContext,
+                Utils.getMapboxAccessToken(this),
+                mapboxNavigationOptions
+        ).also {
+            it.registerRoutesObserver(routeObserver)
+            it.registerRouteProgressObserver(routeProgressObserver)
+        }
+
+        initListeners()
+    }
+
+    override fun onMapReady(mapboxMap: MapboxMap) {
+        mapboxMap.setStyle(Style.MAPBOX_STREETS) { style ->
+            locationEngine = LocationEngineProvider.getBestLocationEngine(this)
+            initLocationComponent(style, mapboxMap)
+            navigationMapboxMap = NavigationMapboxMap(mapView, mapboxMap).also {
+                it.addProgressChangeListener(mapboxNavigation!!)
+                mapInstanceState?.let { state ->
+                    it.restoreFrom(state)
+                }
```

Comment:
There's no necessity to set the camera it has one by default. I think I had to save and restore the state because I had to go between the example and a mock location app in order to test the rerouting and if I didn't restore the state I couldn't see that the app was working.